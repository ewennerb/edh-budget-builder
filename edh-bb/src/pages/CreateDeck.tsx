import React from 'react';
import TextField from '@material-ui/core/TextField';
import firebase from "firebase/app";

interface deckInfo {
  deckName: string;
  deckDescription: string;
  cards: string;
}

class CreateDeck extends React.Component<{ user: firebase.User }> {
  userDocRef: firebase.firestore.DocumentReference;
  constructor(props: Readonly<{ user: firebase.User }>) {
    super(props);
    this.userDocRef = firebase.firestore().collection("users").doc(this.props.user.uid);
    this.state = { value: '' };
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  isValidTitle(title: string) {
    if (title.length > 100) {
      alert("Error: Deck name is too long (must be 100 characters or shorter)");
      return false;
    }
    let hasNonSpaceChar: boolean = false;
    for (var char of title) {
      if (char !== ' ') {
        hasNonSpaceChar = true;
        break;
      }
    }
    if (!hasNonSpaceChar) {
      alert("Error: Your deck title must have at least one alphanumeric character");
      return false;
    }
    return true;
  }

  handleSubmit(event: any, user = this.userDocRef.get()) {
    event.preventDefault();

    let title: string = event.target.deckName.value;
    if (!this.isValidTitle(title)) {
      return;
    }

    console.log("deck is valid");

    firebase.firestore().collection('deck').add({
      deckName: event.target.deckName.value,
      deckDescription: event.target.deckDescription.value,
      deck: [],
      ownerID: user
    })
      .then(function (deckRef) {
        alert("Deck written with ID: " + deckRef.id);
      })

    alert('values input into database: name=' + event.target.deckName.value + ', description=' + event.target.deckDescription.value);
  }

  render() {
    return (
      <form
        onSubmit={this.handleSubmit}
        className="innerForm">
        <h1>[CreateDeck]</h1>
        <TextField
          required
          id="deckName"
          label="Deck Name"
          margin="normal"
        />
        <br></br>
        <TextField
          id="deckDescription"
          placeholder="Deck Description"
          label="Deck Description"
          multiline
          rowsMax="4"
          margin="normal"
        />
        <br></br>
        <br></br>
        <input type="submit" />
      </form>
    );
  }
}

export default CreateDeck;
