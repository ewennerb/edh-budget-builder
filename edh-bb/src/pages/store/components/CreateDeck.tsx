import React, { Component } from 'react';

import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';

import TextField from '@material-ui/core/TextField';

import firebase from "../../../index"


export default class PostDeck extends Component {
  constructor(props: any) {
    super(props);
    this.state = { value: '' };

    this.handleSubmit = this.handleSubmit.bind(this);
  }



  handleSubmit(event: any) {
    event.preventDefault();

    //Add check that deck contains non space chars
    let title: string = event.target.deckName.value;
    let hasNonSpaceChar: boolean = false;
    if (title.length > 100) {
      alert("Error: Your deck name must be shorter than 100 characters");
      return;
    }
    for (var char of title) {
      if (char != ' ') {
        hasNonSpaceChar = true;
      }
    }
    if (!hasNonSpaceChar) {
      alert("Error: Your deck must have at least one non-space character");
      return;
    }


    firebase.firestore().collection('deck').add({
      deckName: event.target.deckName.value,
      deckDescription: event.target.deckDescription.value,
      deck: [1, 2, 4, 5]
    })
      .then(function (deckRef) {
        alert("Deck written with ID: " + deckRef.id);
      })
    alert('values input into database: name=' + event.target.deckName.value + ', description=' + event.target.deckDescription.value);

  }

  render() {
    const classes = useStyles();
    const [values, setValues] = React.useState<State>({
      deckName: "test123",
      deckDescription: "test",
      cards: "test"
    });

    const handleChange = (name: keyof State) => (event: React.ChangeEvent<HTMLInputElement>) => {
      setValues({ ...values, [name]: event.target.value });
    };

    return (
      <form
        onSubmit={this.handleSubmit}
        className="innerForm">
        <h1>[CreateDeck]</h1>
        <TextField
          required
          id="deckName"
          label="Required"
          defaultValue="Deck Name"
          className={classes.textField}
          margin="normal"
          onChange={handleChange('deckName')}
        />
        <br></br>
        <TextField
          id="deckDescription"
          placeholder="Deck Description"
          multiline
          rowsMax="4"
          onChange={handleChange('deckDescription')}
          className={classes.textField}
          margin="normal"
        />
        <br></br>
        <br></br>
        <input type="submit" />
      </form>





    );
  }
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    container: {
      display: 'flex',
      flexWrap: 'wrap',
    },
    textField: {
      marginLeft: theme.spacing(1),
      marginRight: theme.spacing(1),
      width: 1000,
    },
    dense: {
      marginTop: 19,
    },
    menu: {
      width: 200,
    },
  }),
);


interface State {
  deckName: string;
  deckDescription: string;
  cards: string;
}