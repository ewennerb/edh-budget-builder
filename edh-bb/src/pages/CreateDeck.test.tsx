import React from 'react';
import CreateDeck from './CreateDeck';
import firebase from 'firebase/app';
import {render,fireEvent} from '@testing-library/react'
import firebasemock from 'firebase-mock';

jest.mock('firebase/app');
const mockfirestore = new firebasemock.MockFirestore();
mockfirestore.autoFlush(0)
firebase.firestore = (() => mockfirestore) as any;

it("renders", () => {
  
  const testUser: firebase.User = { uid: "testUidAbc123" } as firebase.User;
  var createDeckObj = new CreateDeck({user: testUser});
  var rendered = createDeckObj.render();
  
  expect(rendered).toMatchSnapshot();
})


it("changes text input", async() => {
 
  


  const testUser: firebase.User = { uid: "testUidAbc123" } as firebase.User;


  const { getByRole, getByTestId, getByLabelText,container} =  await render(<CreateDeck user={testUser} />);
 
  const deckName =await container.querySelector("#deckName") as HTMLInputElement;
  const deckDesc =await container.querySelector("#deckDescription") as HTMLInputElement;
  const submit = await getByTestId("submit")
  if(deckName==null){
    console.log("null")
    expect("false").toBeTruthy
  }
  var event = new Event('change');

  await fireEvent.keyDown(deckName, { target: { "value": "testName" } });
  await expect(deckName.value).toBe('testName');
  await fireEvent.change(deckDesc, { target: { "value": "testDesc" } });
  await expect(deckDesc.value).toBe('testDesc');
})


it("addDeckToDatabase", async() => {
 
  const testUser: firebase.User = { uid: "testUidAbc123" } as firebase.User;


  const { getByRole, getByTestId, getByLabelText,container} =  await render(<CreateDeck user={testUser} />);
 
  const deckName =await container.querySelector("#deckName");
  const deckDesc =await container.querySelector("#deckDescription");
  const submit = await getByTestId("submit")
  if(deckName==null){
    console.log("null")
    expect("false").toBeTruthy
  }

  var event = new Event('change');

  await fireEvent.submit(submit, { target: { "deckName": "testName","deckDescription": "testDesc" } });


  var deck = await firebase.firestore().collection('deck').get();

    deck.forEach(deckItem => {
      
      expect(deckItem.data().ownerID).toStrictEqual("testUidAbc123");
      
    })
    
})

