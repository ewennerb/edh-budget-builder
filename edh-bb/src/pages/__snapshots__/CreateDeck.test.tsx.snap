// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<form
  className="innerForm"
  onSubmit={[Function]}
>
  <h1>
    [CreateDeck]
  </h1>
  <ForwardRef(WithStyles)
    id="deckName"
    label="Deck Name"
    margin="normal"
    required={true}
  />
  <br />
  <ForwardRef(WithStyles)
    id="deckDescription"
    label="Deck Description"
    margin="normal"
    multiline={true}
    placeholder="Deck Description"
    rowsMax="4"
  />
  <br />
  <br />
  <button
    data-testid="submit"
  >
    Submit
  </button>
</form>
`;
